### PR TITLE
FEATURE: explicit + crank nicolson implementation

### PR DESCRIPTION
**Pull request by chemardes**
## Main changes

Implementation of explicit (forward-in-time) and Crank-Nicolson method for the 1D heat equation.

## Associated issues 

- Addresses #3

### CODE REVIEW

File: HeatEquationSolver/heat_equation.py [hunk 1]
```diff
@@ -4,57 +4,70 @@
...
-ax.set_xlabel('Length across the rod')
-ax.set_ylabel('Temperature')
-# plotting initial temperature distribution at t = 0
-line, = ax.plot(x, u[0])
-# setting y-axis limits
-ax.set_ylim(bottom=0, top=60)
+def setting_conditions(t, time_step, nodes):
+    # initial and boundary conditions
+    u_initial = 25
+    u_left = 20 * np.sin(np.pi * t) + 25
```
Comment (alphaville) [inline HeatEquationSolver/heat_equation.py:26]: I wouldn't hard-code the initial conditions inside this function

File: HeatEquationSolver/heat_equation.py [hunk 2]
```diff
@@ -4,57 +4,70 @@
...
 
-dx = length / (nodes-1)
-dt_max = 0.5 * (dx**2) / k  # calculating dt using a formula (to ensure stability)
-dt = 0.8 * dt_max
-time_step = int(time/dt)
 
-# creating grid
-x = np.linspace(0, length, nodes)
-t= np.linspace(0, time, time_step)
+def initialise_grid(length, nodes, time, k):
```
Comment (alphaville) [inline HeatEquationSolver/heat_equation.py:12]: Add docstrings

**Comment by chemardes** [comment]
@alphaville pushed my changes there!

File: HeatEquationSolver/heat_equation.py [hunk 3]
```diff
@@ -20,28 +30,82 @@ def initialise_grid(length, nodes, time, k):
...
     return u, u_left, u_right
 
 
-# initializing 2d-matrix to store temperature
 def solve_heat_equation(u, u_left, u_right, time_step, nodes, dt, k, dx):
+    """
+    Solves the heat equation using the explicit scheme
+
+    :param u: 2d-matrix initialised with the initial temperature of the rod
+    :param u_left: array denoting the temperature at the left boundary over time
```
Comment (alphaville) [inline HeatEquationSolver/heat_equation.py:81]: Are `u_left` and `u_right` arrays or functions?

File: HeatEquationSolver/heat_equation.py [hunk 4]
```diff
@@ -20,28 +30,82 @@ def initialise_grid(length, nodes, time, k):
...
 
     return u, u_left, u_right
 
 
-# initializing 2d-matrix to store temperature
 def solve_heat_equation(u, u_left, u_right, time_step, nodes, dt, k, dx):
+    """
+    Solves the heat equation using the explicit scheme
+
+    :param u: 2d-matrix initialised with the initial temperature of the rod
```
Comment (alphaville) [inline HeatEquationSolver/heat_equation.py:80]: If `u` has been initialised already, why do you need to pass `u_left` and `u_right` too?

File: HeatEquationSolver/heat_equation.py (hunk 3 above)
Comment (chemardes) [reply HeatEquationSolver/heat_equation.py:81]: u_left and u_right are arrays consisting of varying temperatures

File: HeatEquationSolver/heat_equation.py (hunk 4 above)
Comment (chemardes) [reply HeatEquationSolver/heat_equation.py:80]: the 'u' matrix im passing into the function only has the initial temp initialised, not the left + right boundary temperatures

Comment (alphaville) [reply HeatEquationSolver/heat_equation.py]: Why not left and right as well? Here's a suggestion: create a function of the form:

```python
grid_params = {"dx": ..., "dt": ...} # you can make this a class later (you can choose a non-uniform grid etc)
solver_params = {"type": "cn"}
left = lambda t: some_function(t)
right = lambda t: some_function(t)
init = lambda x: some_function(x)
sol = solve_heat_equation(kappa, grid_params, solver_params, left, right, init)
# sol is a dictionary/class object with the solution
# if it is a class object, you can do
sol.plot()
# you can do things like
mean_temp = sol.mean_temperature()
# and you can also compare two solutions
error = sol1.diff(sol2)
```

That said, your don't need to do this now. We can merge it now and update it later. Your call.

**Comment by chemardes** [comment]
@alphaville changes pushed - pls ignore heat_equation_cn.py, that's just the original explicit method implementation copied!

File: HeatEquationSolver/heat_equation.py (hunk 4 above)
Comment (chemardes) [reply HeatEquationSolver/heat_equation.py]: will look into it, thank you !

File: HeatEquationSolver/heat_equation_cn.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    lhs = build_tridiagonal_matrix(a, b, c, nodes-2)
+    rhs = np.zeros(nodes-2)
+
+    for tau in range(0, time_step - 1):
+        rhs[0] = alpha * (u[tau, 0] + u[tau+1, 0]) + (1 - 2 * alpha) * u[tau, 1] + alpha * u[tau, 2]
+
+        for i in range(1, nodes-3):
+            rhs[i] = alpha * u[tau, i] + (1 - 2 * alpha) * u[tau, i+1] + alpha * u[tau, i+2]
+
+        rhs[-1] = alpha * (u[tau, -1] + u[tau+1, -1]) + (1 - 2 * alpha) * u[tau, -2] + alpha * u[tau, -3]
```
Comment (chemardes) [inline HeatEquationSolver/heat_equation_cn.py:82]: i feel like there's something wrong with this

File: HeatEquationSolver/heat_equation.py (hunk 4 above)
Comment (alphaville) [reply HeatEquationSolver/heat_equation.py]: @chemardes what did you decide?

File: HeatEquationSolver/heat_equation_cn.py (hunk 5 above)
Comment (alphaville) [reply HeatEquationSolver/heat_equation_cn.py:93]: You should check whether the explicit method and the CN method give you similar solutions.

File: HeatEquationSolver/heat_equation.py (hunk 4 above)
Comment (chemardes) [reply HeatEquationSolver/heat_equation.py]: i think i'll use your implementation - makes things more structured!

File: HeatEquationSolver/heat_equation_cn.py (hunk 5 above)
Comment (chemardes) [reply HeatEquationSolver/heat_equation_cn.py:93]: it does with slight differences !

**Comment by debdal6** [comment]
Thats all you need for the 3d surface plot
`# Visualization
# Create a meshgrid for plotting
X, Y = numpy.meshgrid(timeDomain, xDomain)

# Plot the 3D surface
fig = pyPlot.figure(figsize=(10, 7))
ax = fig.add_subplot(111, projection='3d')
surface = ax.plot_surface(X, Y, tempMatrix, cmap='inferno')

# Set labels and title
ax.set_xlabel('Time')
ax.set_ylabel('Position along the rod (x)')
ax.set_zlabel('Temperature')
ax.set_title('1D Heat Diffusion Simulation')

pyPlot.show()`

**Comment by chemardes** [comment]
@debdal6
' X, Y = numpy.meshgrid(timeDomain, xDomain) '

why not numpy.meshgrid(xDomain, timeDomain)?

**Comment by debdal6** [comment]
> @debdal6 ' X, Y = numpy.meshgrid(timeDomain, xDomain) '
> 
> why not numpy.meshgrid(xDomain, timeDomain)?

my matrix is inverted, basically the syntax is `numpy.meshgrid(x,y)`

**Comment by chemardes** [comment]
> > @debdal6 ' X, Y = numpy.meshgrid(timeDomain, xDomain) '
> > why not numpy.meshgrid(xDomain, timeDomain)?
> 
> my matrix is inverted, basically the syntax is `numpy.meshgrid(x,y)`

gotcha i just realised lol

**Comment by alphaville** [comment]
@chemardes please let me know when you have updated your code in this PR. Let's try to merge this soon.

**Comment by chemardes** [comment]
> @chemardes please let me know when you have updated your code in this PR. Let's try to merge this soon.

@alphaville will do by eod!

File: playground/heat_equation.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        line.set_ydata(u[frame])
+        ax.set_title(f'Heat Distribution at time = {frame * dt:.2f}s')
+
+        return line,
+
+    ani = FuncAnimation(fig, animate, frames=range(0, time_step, time_step // 100), blit=False)
+    plt.show()
+
+
+def main():
```
Comment (alphaville) [inline playground/heat_equation.py:97]: It's a little strange that `initial_temp` is not a function of $x$. In any case, we can merge this and you can update it in a future version.

File: playground/heat_equation_cn.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    :param length: length of the rod
+    :param nodes: number of spatial nodes across the rod
+    :param time: the total time
+    :param k: heat diffusivity constant
+    :param initial_temp: initial temperature of the rod at t=0
+    :return: - u: the solved 2d matrix consisting the temperature distribution over time
+             - dt: time step size (used for plotting)
+             - time_step: no. of time steps for the simulation (used for plotting)
+    """
+    dx = length / (nodes-1)
```
Comment (alphaville) [inline playground/heat_equation_cn.py:70]: Lines 71-72 are unnecessary. The Crank-Nicolson method is **unconditionally stable**.

**Review by alphaville — CHANGES_REQUESTED** [review]
@chemardes Just a minor comment on the CN method. You can merge this afterwards.

**Comment by chemardes** [comment]
> @chemardes Just a minor comment on the CN method. You can merge this afterwards.

fixed

**Comment by alphaville** [comment]
> > @chemardes Just a minor comment on the CN method. You can merge this afterwards.
> 
> fixed

@chemardes, feel free to merge

File: playground/heat_equation.py (hunk 6 above)
Comment (alphaville) [reply playground/heat_equation.py:97]: If you like you can merge this now and we'll revisit this in a future PR.

**Pull request merged**